conn.go

```diff
@@ -476,11 +490,91 @@ func (cn *conn) gname() string {
 	return strconv.FormatInt(int64(cn.namei), 10)
 }
 
-func (cn *conn) simpleExec(q string) (res driver.Result, commandTag string, err error) {
-	b := cn.writeBuf('Q')
-	b.string(q)
-	cn.send(b)
+func (cn *conn) awaitSynchronizationPoint() {
+	t, r := cn.recv1()
+	switch t {
+	case 'Z':
+		cn.processReadyForQuery(r)
+		return
+	default:
+		cn.bad = true
+		errorf("unexpected message %q while waiting for synchronization point")
+	}
+}
+
+func (cn *conn) readParseResponse() {
+	t, r := cn.recv1()
+	switch t {
+	case '1':
+		return
+	case 'E':
+		err := parseError(r)
+		cn.awaitSynchronizationPoint()
+		panic(err)
+	default:
+		cn.bad = true
+		errorf("unexpected Parse response %q", t)
+	}
+}
+
+func (cn *conn) readBindResponse() {
+	t, r := cn.recv1()
+	switch t {
+	case '2':
+		return
+	case 'E':
+		err := parseError(r)
+		cn.awaitSynchronizationPoint()
+		panic(err)
+	default:
+		cn.bad = true
+		errorf("unexpected Bind response %q", t)
+	}
+}
+
+func (cn *conn) readStatementDescribeResponse() (paramTyps []oid.Oid, cols []string, rowTyps []oid.Oid) {
+	t, r := cn.recv1()
+	switch t {
+	case 't':
+		nparams := r.int16()
+		paramTyps = make([]oid.Oid, nparams)
+		for i := range paramTyps {
+			paramTyps[i] = r.oid()
+		}
+	case 'E':
+		err := parseError(r)
+		cn.awaitSynchronizationPoint()
+		panic(err)
+	default:
+		cn.bad = true
+		errorf("unexpected Describe response %q", t)
+	}
+
+	// Cheat a bit since the result should be exactly the same as when
+	// describing a portal.
```

Comment:
Not true. If I understand correctly,

- A RowDescription after Query or Bind includes the text/binary format indicators; in both cases a portal exists.
- Otherwise, we must use/expect formats as specified in Bind.